### PR TITLE
add option to simulate emitters

### PR DESCRIPTION
**Pull request by cg-laser**
This pull requests will add the option to simulate emitter with arbitrary polarization to NuRadioMC. This is useful to simulate calibration measurments. 
Ilya summarized the problem with the current code very well and I quote him below:

> Hi Christian,
> While discussing implementation details between me and Pawan, we came to realization of a fundamental difficulty of implementing a radio emitter as simply another plug-and-play option alongside with various askaryan shower models in NuRadioMC.
> 
>   Let me run by you my understanding of how simulation happens:
> 
> =========
> 
> 1. The electric field emitted by a source is assumed to be radial. All models in askaryan.py return a 1D E-field vector sampled in time, which is taken to be the “theta” component at this point. This is seen in here, in the get_time_trace function.
> 
> https://github.com/nu-radio/NuRadioMC/blob/857991c8e06db00196e222606e99654cc76b2f90/NuRadioMC/SignalGen/askaryan.py#L14-L17
> 
> In Pawan’s code, we have get_time_trace implemented for a radio emitter of chosen design that is fed a chosen V_src(time) pattern. We produce E(time) trace, where E is the theta polarization component (p-mode) while the phi polarization component is (s-mode) is ignored for the time being.
> 
> 2. What is actually used for event simulation is the E-field as a function of frequency, the get_frequency_spectrum of askaryan.py:
> 
> https://github.com/nu-radio/NuRadioMC/blob/857991c8e06db00196e222606e99654cc76b2f90/NuRadioMC/SignalGen/askaryan.py#L122
> 
> which simply calls the get_time_trace and makes an fft spectrum out of that.
> 
> 3. The get_time_trace is used from the simulation.py file:
> 
> https://github.com/nu-radio/NuRadioMC/blob/857991c8e06db00196e222606e99654cc76b2f90/NuRadioMC/simulation/simulation.py#L698
> 
> At this point, the electric field is 1D array E(f) representing nominally “theta” polarization. 
> 
> 4. Then, the electric field vector, originally nominally “theta”, gets rotated appropriate to the geometry of the given event in this line:
> 
> https://github.com/nu-radio/NuRadioMC/blob/857991c8e06db00196e222606e99654cc76b2f90/NuRadioMC/simulation/simulation.py#L726
> 
> which, as far as I understand, rotates the original “theta” vector appropriately so that the E-field vector is in the plane that goes through the neutrino line-of-flight, the receiver antenna location, and is directed away from the neutrino line-of-flight in that plane (assuming I am reading the code correctly). Thus, the E-field vector now has both theta and phi transverse polarizations that depend on the neutrino line-of-flight and receiver position.
> 
> ==========
> 
> This implementation, assuming I am right, is presenting us with difficulties. What we would ideally do is we would generate both theta and phi components of the electric field for a generic radio emitter, and we would disable the rotation of the polarization vector for such an emitter that assumes radial polarization direction. As far as I can see, it would be non-trivial, with get_time_trace and get_frequency_spectrum returning 2D array covering both transverse polarizations (with the second polarization being zero for neutrino showers), and then simulation.py handing that appropriately, rotating or not rotating the E-field vector depending on whether the source is a shower or a generic emitter. What solutions do you see given your knowledge?
> 
> On the shorter term, it is still possible to work with VPol emitters, it seems to me, setting the dummy “neutrino” direction to point anywhere as long as it lies in a vertical plane that goes through the receiver and the emitter locations. For this to work for a station with multiple receiver locations, one would need to point the dummy neutrino direction straight up or straight down. This would assume that pick-up on the other polarization is negligible.
> 
> However, an HPol transmitter, seems to me, would be not possible to simulate this way, both because with the difficulty to define the “dummy” neutrino direction and the fact that HPol (I think) emits both non-negligible theta and phi components to start with.
> 
> Please let me know if we are misunderstanding something or if you think non-trivial changes to askaryan.py and simulation.py that would accumulate handling two transverse polarizations without assuming radial direction with respect to a shower axis could be possible.
> 
> Thanks,
> Ilya

I think the best option is to add another config parameter that specifies if the signal source is a shower or an emitter (with the potential to add more classes in the future). The main difference is that the emitter model directly returns a 3-dim electric field. I created a new file called "emitter.py" which is similar to "askaryan.py" and has the same interface but returns a 3d vector.  This file can then store a variety of different emitters. Please have a look at the code and tell me what you think. 

I have a few questions myself
* should we better integrate emitter.py into askrayan.py? This would help to keep the two interfaces in sync but it would be messier. Another kwarg would then specify if a 1d or 3d trace is returned. 
* should we make the emitters themselves modular? I.e. emitter.py would only contain the voltage output of the pulser. Then we specify the emitting antenna and orientation and NuRadioMC automatically calculates the resulting efield. This can be achieved by a new NuRadioReco module that convolves the voltage with the antenna response. I think this would be the nicest solution. The exact signal chain (potentially there is another bandpass filter or amp in the setup), can be specified equivalent to the `def _detector_simulation(self):` method, lets call it `def _emitter_simulation(self)`. The only thing I'm not sure about is where to specify the emitter antenna configuration. At first maybe just manually in `_emitter_simulation`?

### CODE REVIEW

**Comment by ikrav** [comment]
This looks nice and simple.

I am more of a C++ person. In C++ I would want to create an abstract base class Emitter with derived classes NeutrinoShowerEmitter and PulserEmitter, with expectation of CosmicRayShowerCoreEmitter or more. The base class would define the interface. Then I would define a base class virtual function for get_time_trace and overload it in all derived classes with all functions returning 3d trace (passing maybe the launch vector into the Emitter). But this is python, and I do not know good practices in python well enough. I've heard that python does have abstract classes and interfaces since recently. I do agree that it would be nice to enforce a common interface without having multiple places to edit in case of changes.

Perhaps we should just give this implementation a go.

File: NuRadioMC/SignalGen/emitter.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
+# -*- coding: utf-8 -*-
+import numpy as np
+from NuRadioReco.utilities import units, fft
+from NuRadioMC.SignalGen import parametrizations as par
```
Comment (ikrav) [inline NuRadioMC/SignalGen/emitter.py:4]: Is this line needed?

File: NuRadioMC/SignalGen/emitter.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+import logging
+logger = logging.getLogger("SignalGen.askaryan")
+
+
+def set_log_level(level):
+    logger.setLevel(level)
+    par.set_log_level(level)
```
Comment (ikrav) [inline NuRadioMC/SignalGen/emitter.py:11]: Is this line needed?

**Review by ikrav — COMMENTED** [review]
Overall looks good to me, just two lines in the emitter.py appear to me unneeded.

File: NuRadioMC/SignalGen/emitter.py (hunk 2 above)
Comment (cg-laser) [reply NuRadioMC/SignalGen/emitter.py:11]: good point. These lines should be removed.

**Comment by cg-laser** [comment]
I agree, in C++ we would write a base class to ensure the same interface. As far as I know, this won't work in python because you always overwrite the functions any way you want. Let's see what others think @shallmann @christophwelling ?

**Comment by ikrav** [comment]
Christian, there is one thing that Pawan noticed. Presently, in simulation.py the function call of get_frequency_spectrum() for the emitter case does not contain the field "model" while the function implementation in the emitter.py does have it.
The call:
   https://github.com/nu-radio/NuRadioMC/blob/feature/emitter/NuRadioMC/simulation/simulation.py#L733
the implementation:
   https://github.com/nu-radio/NuRadioMC/blob/feature/emitter/NuRadioMC/SignalGen/emitter.py#L74
Generally, we need to pass into the emitter module some number of parameters, at the very least:
   - the V_src(t) pattern (such as HVSP, IDL pulsers, CW train, etc), which could be the "model"
   - the amplitude of the V_src(t), which could be the existing "energy" field
   - the transmitting antenna type (such as bicone_v8)
 Possibly other parameters would also be needed in the future. Should we at least add the missing "model" field and additionally introduce "antenna_type"?

     I also wonder if it makes sense to define in the yaml config for event generation a section "emitter" with fields "model", "antenna_type", "amplitude", etc. Then do in simulation.py something like 
    emitterConifig = self._cfg['emitter']
where emitterConfig woud only contain the emitter description fields from the yaml config. The emitterConfig can then be passed into the get_frequency_spectrum and then into the get_time_trace in the emitter.py. (I do not know if this can be done exactly as above, or if there are any reasons why it shouldn't be done). The advantage of this would be in the fact that there will be no need to change siulation.py or even emitter.py while a specific implementation of an emitter is being developed. The changes would be confined to the config.yaml and the separate .py file for a specific emitter implementation (I expect we won't have explicit calculations of anything in the emitter.py itself, but will have that in a separate .py file/class).

**Review by christophwelling — APPROVED** [review]
Looks good to me.

About writing a mother class for the emittters: I think the python way to do it would be to write an emitter class that has the get_trace() and get_frequency_spectrum(), which define the interface. Internally, these methods would call _get_trace() and _get_frequency(), which would be overridden by the child classes.

In this case, that seems a bit like overkill, so I am fine with leaving it as it is for now.

**Comment by cg-laser** [comment]
@ikrav I started implementing what we discussed on Thursday. I created a new test for it so you directly have a working example which I also added to our automated testing routine. So far the emitter code (covolving the voltage output with the emitting antenna response ) is just in the main simulation.py code. We should probably refactor this later and put it into a separate function. But for now, we can also just leave it there to test it. You still need to implement the correct way of convolving the antenna. Let me know what you think of the new interface. 

One thing that is not very nice is that the input file still needs to contain fields like `shower_energy` although it is not used anywhere. I'll think about how we can clean up the code. It does not impact the functionality of the code, just its beauty ;-)

**Comment by ikrav** [comment]
@cg-laser This looks good to me in general from just looking in the code, Pawan and I should run it to get a better feel for it. My immediate comment was, as you point out, that having actual calculations with VEL and stuff in simulation.py is not appropriate, but we can start with this and move it out later. Having a test setup that you implemented is very helpful.

**Comment by pgiri4** [comment]
Thanks, Christian.  tried this one for spherical and square signals and looks good for now. We can try loading VEL outside the simulation.py . Once I find it working with all models then I will let you know .














[https://ucomm.unl.edu/images/email-signature/Nebraska_N_RGB_small_on_white.gif]

Pawan Giri

Graduate Teaching Asst

University of Nebraska–Lincoln

Physics & Astronomy

JH-253

________________________________

**Comment by cg-laser** [comment]
This PR was continued on #342

**Pull request closed without merging**